File: src/net/dinkla/raytracer/objects/acceleration/kdtree/SpatialMedianBuilder.java
package net.dinkla.raytracer.objects.acceleration.kdtree;

import net.dinkla.raytracer.math.BBox;
import net.dinkla.raytracer.math.Point3D;
import net.dinkla.raytracer.math.Vector3D;
import net.dinkla.raytracer.objects.GeometricObject;
import net.dinkla.raytracer.math.Axis;
import net.dinkla.raytracer.utilities.Counter;
import org.apache.log4j.Logger;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by IntelliJ IDEA.
 * User: jorndinkla
 * Date: 11.06.2010
 * Time: 11:46:03
 * To change this template use File | Settings | File Templates.
 */
public class SpatialMedianBuilder implements IKDTreeBuilder {

    static final Logger LOGGER = Logger.getLogger(SpatialMedianBuilder.class);

    public int maxDepth = 15;
    public int minChildren = 4;

    public AbstractNode build(KDTree tree, BBox voxel) {
        return build(tree.getObjects(), tree.getBoundingBox(), 0);
    }

    public AbstractNode build(List<GeometricObject> objects, BBox voxel, int depth) {

        Counter.count("KDtree.build");

        AbstractNode node = null; //new Leaf(objects);
        BBox voxelL = null;
        BBox voxelR = null;

        if (objects.size() < minChildren || depth >= maxDepth) {
            Counter.count("KDtree.build.leaf");
            node = new Leaf(objects);
            return node;
        }

        Counter.count("KDtree.build.node");

        Vector3D half = voxel.getQ().minus(voxel.getP()).mult(0.5);
        Point3D mid = voxel.getP().plus(half);

        Double split = null;

        List<GeometricObject> objectsL = new ArrayList<GeometricObject>();
        List<GeometricObject> objectsR = new ArrayList<GeometricObject>();

        if (depth % 3 == 0) {
            // x
            split = mid.getX();

            Point3D q1 = new Point3D(mid.getX(), voxel.getQ().getY(), voxel.getQ().getZ());
            voxelL = new BBox(voxel.getP(), q1);

            Point3D p2 = new Point3D(mid.getX(), voxel.getP().getY(), voxel.getP().getZ());
            voxelR = new BBox(p2, voxel.getQ());

            for (GeometricObject object : objects) {
                BBox bbox = object.getBoundingBox();
                if (bbox.getP().getX() <= split) {
                    objectsL.add(object);
                }
                if (bbox.getQ().getX() >= split) {
                    objectsR.add(object);
                }
            }

        } else if (depth % 3 == 1) {
            // y
            split = mid.getY();

            Point3D q1 = new Point3D(voxel.getQ().getX(), mid.getY(), voxel.getQ().getZ());
            voxelL = new BBox(voxel.getP(), q1);

            Point3D p2 = new Point3D(voxel.getP().getX(), mid.getY(), voxel.getP().getZ());
            voxelR = new BBox(p2, voxel.getQ());

            for (GeometricObject object : objects) {
                BBox bbox = object.getBoundingBox();
                if (bbox.getP().getY() <= split) {
                    objectsL.add(object);
                }
                if (bbox.getQ().getY() >= split) {
                    objectsR.add(object);
                }
            }
        } else if (depth % 3 == 2) {
            // z
            split = mid.getZ();

            Point3D q1 = new Point3D(voxel.getQ().getX(), voxel.getQ().getY(), mid.getZ());
            voxelL = new BBox(voxel.getP(), q1);

            Point3D p2 = new Point3D(voxel.getP().getX(), voxel.getP().getY(), mid.getZ());
            voxelR = new BBox(p2, voxel.getQ());

            for (GeometricObject object : objects) {
                BBox bbox = object.getBoundingBox();
                if (bbox.getP().getZ() <= split) {
                    objectsL.add(object);
                }
                if (bbox.getQ().getZ() >= split) {
                    objectsR.add(object);
                }
            }
        }

        LOGGER.info("Splitting " + objects.size() + " objects into " + objectsL.size() + " and " + objectsR.size() + " objects at " + split + " with depth " + depth);
        AbstractNode left = build(objectsL, voxelL, depth + 1);
        AbstractNode right = build(objectsR, voxelR, depth + 1);

        node = new InnerNode(left, right, voxel, split, Axis.Companion.fromInt(depth % 3));

        return node;
    }

    public int getMaxDepth() {
        return maxDepth;
    }

    public void setMaxDepth(int maxDepth) {
        this.maxDepth = maxDepth;
    }
}

File: src/commonMain/kotlin/net/dinkla/raytracer/objects/acceleration/kdtree/builder/SpatialMedianBuilder.kt
package net.dinkla.raytracer.objects.acceleration.kdtree.builder

import net.dinkla.raytracer.math.Axis
import net.dinkla.raytracer.math.BBox
import net.dinkla.raytracer.math.Point3D
import net.dinkla.raytracer.objects.IGeometricObject
import net.dinkla.raytracer.objects.acceleration.kdtree.InnerNode
import net.dinkla.raytracer.objects.acceleration.kdtree.KDTree
import net.dinkla.raytracer.objects.acceleration.kdtree.Leaf
import net.dinkla.raytracer.objects.acceleration.kdtree.Node
import net.dinkla.raytracer.utilities.Counter
import net.dinkla.raytracer.utilities.Logger

/**
 * The default [TreeBuilder]: splits each voxel at its **spatial median** — the geometric midpoint of
 * the current axis — cycling axes by depth (x, y, z, x, …). An object goes left when its lower bound
 * is `<= split` and right when its upper bound is `>= split`, so objects straddling the plane are
 * duplicated into both children. Recursion stops at [maxDepth] or fewer than `minChildren` objects.
 */
class SpatialMedianBuilder : TreeBuilder {
    override var maxDepth = 15
    private var minChildren = 4

    override fun build(
        tree: KDTree,
        voxel: BBox,
    ): Node = build(tree.objects, tree.boundingBox, 0)

    /** Recursively builds the subtree for [objects] within [voxel] at the given [depth]; see the class doc. */
    fun build(
        objects: List<IGeometricObject>,
        voxel: BBox?,
        depth: Int,
    ): Node {
        Counter.count("KDtree.build")

        var node: Node? = null // new Leaf(objects);
        var voxelL: BBox? = null
        var voxelR: BBox? = null

        if (objects.size < minChildren || depth >= maxDepth) {
            Counter.count("KDtree.build.leaf")
            node = Leaf(objects)
            return node
        }

        Counter.count("KDtree.build.node")

        requireNotNull(voxel) { "voxel must be non-null for an inner node at depth $depth" }
        val half = voxel.q.minus(voxel.p).times(0.5)
        val mid = voxel.p.plus(half)

        var split: Double? = null

        val objectsL = ArrayList<IGeometricObject>()
        val objectsR = ArrayList<IGeometricObject>()

        if (depth % 3 == 0) { // x
            split = mid.x

            val q1 = Point3D(mid.x, voxel.q.y, voxel.q.z)
            voxelL = BBox(voxel.p, q1)

            val p2 = Point3D(mid.x, voxel.p.y, voxel.p.z)
            voxelR = BBox(p2, voxel.q)

            for (geometricObject in objects) {
                val bbox = geometricObject.boundingBox
                if (bbox.p.x <= split) {
                    objectsL.add(geometricObject)
                }
                if (bbox.q.x >= split) {
                    objectsR.add(geometricObject)
                }
            }
        } else if (depth % 3 == 1) { // y
            split = mid.y

            val q1 = Point3D(voxel.q.x, mid.y, voxel.q.z)
            voxelL = BBox(voxel.p, q1)

            val p2 = Point3D(voxel.p.x, mid.y, voxel.p.z)
            voxelR = BBox(p2, voxel.q)

            for (geometricObject in objects) {
                val bbox = geometricObject.boundingBox
                if (bbox.p.y <= split) {
                    objectsL.add(geometricObject)
                }
                if (bbox.q.y >= split) {
                    objectsR.add(geometricObject)
                }
            }
        } else if (depth % 3 == 2) { // z
            split = mid.z

            val q1 = Point3D(voxel.q.x, voxel.q.y, mid.z)
            voxelL = BBox(voxel.p, q1)

            val p2 = Point3D(voxel.p.x, voxel.p.y, mid.z)
            voxelR = BBox(p2, voxel.q)

            for (geometricObject in objects) {
                val bbox = geometricObject.boundingBox
                if (bbox.p.z <= split) {
                    objectsL.add(geometricObject)
                }
                if (bbox.q.z >= split) {
                    objectsR.add(geometricObject)
                }
            }
        }

        Logger.info(
            "Splitting " + objects.size + " objects into " + objectsL.size + " and " +
                objectsR.size + " objects at " + split + " with depth " + depth,
        )
        val splitValue = requireNotNull(split) { "split must be computed for an inner node at depth $depth" }
        val left = build(objectsL, voxelL, depth + 1)
        val right = build(objectsR, voxelR, depth + 1)
        return InnerNode(left, right, voxel, splitValue, Axis.fromInt(depth))
    }
}
